keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,645 @@
+   (`kusttarget.accumulateResources`):
+  * Adds files to the RA directly (`accumulateFile`).
+  * Recursively processes Kustomizations, each starting with a new empty RA,
+    and with the result merged in to the RA of the parent
+    (`accumulateDirectory` and `resaccumulator.MergeAccumulator`).
+2. Adds/merges its CRDs, generators, transformers, and variables to the RA.
+  * These modifications can only be applied to entities within the RA - it is
+    an error otherwise. This means that sibling Kustomizations are entirely
+    independent of each other.
+  * Kustomizations with the same base will cause an error when the resources
+    are accumulated, unless their GVKN is altered.
+
+`Components` are processed between step 1 and 2, and in a similar manner to
+Kustomizations, but instead of starting with an empty RA, they take the RA from
+their parent. The component is therefore able to modify everything within the
+RA, which contains all of the parent's resources and the result of
+processing all of the earlier components in the `components:` list. Nested
+components work in the same way - ownership of the parent's RA is simply
+passed down.
+
+We implement this by changing the signature of `kusttarget.accumulateResources`
+to return a pointer to the `ResourceAccumulator` that the parent should use,
+and giving child components their parent's RA instead of an empty RA. This
+minimizes the amount of code changed to simplify review, but a slightly more
+extensive refactoring is recommended.
+
+Additionaly, we add some checks to ensure that Kustomizations and normal
+files are not added to the `components:` list and Components are not added to
+`resources:`.
+
+Also, note that we do not change the code that loads the contents of
+Kustomizations, so it should be possible to load Components from any place
+where one expects to load a Kustomization.
+
+Finally, if a new `components:` field is not desired, a variation of the
+implementation could be to allow Components to be added to the `resources:`
+list. In this case:
+
+* The basic implementation of components doesn't need to change.
+* It is unclear whether directories in the `resources:` list are Kustomizations
+  or Components. A naming convention could be used (e.g. store components in a
+  `components/` directory in the same way as overlay Kustomizations are often
+  stored in `overlays/`) but this is less explicit than using a separate
+  `components:` list.
+* The semantics of the `resources:` field will change and order will matter.
+
+### Test Plan
+
+Add unit tests for all the points raised in the "Design details" section.
+
+### Graduation Criteria
+
+#### Alpha -> Beta
+
+- [ ] Implement the necessary functionality for this feature
+- [ ] Write the appropriate unit tests, and make the `TestCompositeDiamond`
+      test, which was originally written to show that the composition feature
+      was not supported, pass
+- [x] Add a user story in Kustomize's examples
```

Comment:
point to merged PR, i.e.:

https://github.com/kubernetes-sigs/kustomize/pull/2438

Reply:
Done.